### PR TITLE
Provide samples for Iterables (KT-20357)

### PR DESCRIPTION
**Pull request by tommykw**
https://youtrack.jetbrains.com/issue/KT-20357

### CODE REVIEW

File: libraries/stdlib/samples/test/samples/collections/iterables.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package samples.collections
+
+import samples.*
+import kotlin.test.*
+
+class Iterables {
+
+    @Sample
+    fun iterable() {
+        val iterator = Sequence { arrayOf(1, 2, 3).iterator() }.iterator()
```
Comment (ilya-g) [inline libraries/stdlib/samples/test/samples/collections/iterables.kt]: This sample doesn't use the function it is attached to.
The function may be useful for turning into `Iterable` something that has iterator or is an `Iterator`, but isn't `Iterable` itself, for example `buildIterator { }`

File: libraries/stdlib/samples/test/samples/collections/iterables.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import samples.*
+import kotlin.coroutines.experimental.buildIterator
+import kotlin.test.*
+
+class Iterables {
+
+    @Sample
+    fun iterable() {
+        val iterator = buildIterator { yieldAll(listOf(1, 2, 3)) }
```
Comment (ilya-g) [inline libraries/stdlib/samples/test/samples/collections/iterables.kt]: This sample still doesn't call `Iterable` function for which it is provided.

**Comment by tommykw** [comment]
Thanks! Could you check this PR, again?

File: libraries/stdlib/samples/test/samples/collections/iterables.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import samples.*
+import kotlin.coroutines.experimental.buildIterator
+import kotlin.test.*
+
+class Iterables {
+
+    @Sample
+    fun iterable() {
+        val iterator = Iterable { buildIterator { yieldAll(listOf(1, 2, 3)) } }.iterator()
```
Comment (ilya-g) [inline libraries/stdlib/samples/test/samples/collections/iterables.kt]: Hmm, now it's confusing why do we get an iterator immediately from that `Iterable`: if one wanted to have an iterator, he could use directly that one returned by `buildIterator`.

Usually we want to wrap something in `Iterable` to pass it where `Iterable` is expected. For example there are a lot of extension functions taking `Iterable` as a receiver: `map`, `filter` and so on.

Comment (ilya-g) [reply libraries/stdlib/samples/test/samples/collections/iterables.kt]: Also list of consecutive numbers here can be replaced with a range `1..3`

Comment (tommykw) [reply libraries/stdlib/samples/test/samples/collections/iterables.kt]: Thanks, fixed.

File: libraries/stdlib/samples/test/samples/collections/iterables.kt [hunk 4]
```diff
@@ -24,7 +24,13 @@ class Iterables {
 
     @Sample
     fun iterable() {
-        val iterator = Iterable { buildIterator { yieldAll(listOf(1, 2, 3)) } }.iterator()
+        val iterator = buildIterator {
```
Comment (ilya-g) [inline libraries/stdlib/samples/test/samples/collections/iterables.kt]: Sorry, I don't see `Iterable` function call here again. Have you forgotten for which function this sample is? :)

What I've meant is something like:
```
val someIterable = Iterable { buildIterator { ... } }

// now you can call extensions on the Iterable
val result: List<String> = someIterable.mapIndexed { index, value -> "$index => $value" }
assertPrints(result, "...")
```

Comment (tommykw) [reply libraries/stdlib/samples/test/samples/collections/iterables.kt]: Really sorry!! I will try to fix samples.

Comment (tommykw) [reply libraries/stdlib/samples/test/samples/collections/iterables.kt]: Thanks, fixed.

**Comment by ilya-g** [comment]
I've merged it manually with some improvements, thanks.

**Pull request closed without merging**